test/unit/org/apache/cassandra/cql3/validation/operations/SelectMultiColumnRelationTest.java

```diff
@@ -1930,4 +1930,38 @@ public void testInvalidColumnNames() throws Throwable
         assertInvalidMessage("Undefined column name e", "SELECT c AS e FROM %s WHERE (b, e) IN ((0, 1), (2, 4))");
         assertInvalidMessage("Undefined column name e", "SELECT c AS e FROM %s WHERE (b, e) > (0, 1) and b <= 2");
     }
+
+    @Test
+    public void testInRestrictionsWithAllowFiltering() throws Throwable
+    {
+        createTable("CREATE TABLE %s (pk int, c1 text, c2 int, v int, primary key(pk, c1, c2))");
+        execute("INSERT INTO %s (pk, c1, c2, v) values (?, ?, ?, ?)", 1, "0", 0, 3);
+        execute("INSERT INTO %s (pk, c1, c2, v) values (?, ?, ?, ?)", 1, "1", 0, 4);
+        execute("INSERT INTO %s (pk, c1, c2, v) values (?, ?, ?, ?)", 1, "1", 1, 5);
+        execute("INSERT INTO %s (pk, c1, c2, v) values (?, ?, ?, ?)", 1, "2", 1, 6);
+        execute("INSERT INTO %s (pk, c1, c2, v) values (?, ?, ?, ?)", 1, "2", 2, 7);
+
+        assertRows(execute("SELECT * FROM %s WHERE (c2) IN ((?), (?)) ALLOW FILTERING", 1, 3),
+                   row(1, "1", 1, 5),
+                   row(1, "2", 1, 6));
+
+        assertRows(execute("SELECT * FROM %s WHERE c2 IN (?, ?) ALLOW FILTERING", 1, 3),
+                   row(1, "1", 1, 5),
+                   row(1, "2", 1, 6));
+    }
+
+    @Test
+    public void testInRestrictionsWithIndex() throws Throwable
+    {
+        createTable("CREATE TABLE %s (pk int, c1 text, c2 int, c3 int, v int, primary key(pk, c1, c2, c3))");
+        createIndex("CREATE INDEX ON %s (c3)");
```

Comment:
The test produces the same `"IN restrictions are not supported on indexed columns"` error if we don't create any index. Indeed, the index is never selected because of the rejection of the `IN` operator by `Index#supportsExpression`, and it's `ClusteringColumnRestrictions#handleInFilter` what leads us to the code raising the exception. 

If we can't process this type of multicolumn restrictions in the row filter maybe the error message should be  something like `"Multicolumn IN filters are not supported"`, or something like that. We should also add an equivalent test without the index.

I guess that at some point CEP-7 will add support for OR conditions in the row filter, I think that then we'll be able to revisit this. wdyt?

Reply:
I was surprised to hit the error. I did not realize that I had kept the `ALLOW FILTERING`. I should have looked deeper.

The error message is not making sense at all. I did not changed it by laziness. So +1 to the change.